### DISCUSSION TITLE
좋은 객체란 무엇이라 생각하나요?

### DISCUSSION THREAD

**Discussion by seunghye218** [2024-woowacourse-study/level-interview#6] [OOP]
.

**Comment by alstn113** [comment]
"좋은 객체"는 여러 측면에서 정의될 수 있지만, 일반적으로 다음과 같은 특성을 가지는 객체라고 생각된다.

1. 응집도: 객체는 한 가지 목적 또는 책임을 가지고 있어야 합니다. 객체가 자신의 책임을 명확하게 이해하고 그에 맞는 동작을 수행하는 것이 중요합니다.
2. 결합도: 객체 간의 결합도가 낮아야 합니다. 다시 말해, 객체는 다른 객체에 대한 의존성을 최소화해야 하며, 이는 객체 간의 유연한 상호작용을 가능하게 합니다.
3. 재사용성: 좋은 객체는 다른 곳에서도 재사용될 수 있는 형태여야 합니다. 즉 , 객체는 재사용 가능한 기능을 캡슐화하고 있어야 합니다.
4. 가독성: 객체의 코드는 가독성이 높아야 합니다. 다른 프로그래머가 쉽게 이해하고 수정할 수 있도록 코드를 작성해야 합니다.
5. 테스트 용이성: 좋은 객체는 테스트하기 쉬워야 합니다. 객체의 상태와 동작을 테스트하는데 필요한 메서드를 제공하고, 테스트에 용이한 코드를 작성해야 합니다.
6. 단일 책임 원칙: 객체는 변경의 이유가 단 하나여야 합니다. 이는 객체가 변경될 가능성이 있는 이유를 최소화하고, 유지보수성을 향상시킵니다.
7. 데이터 은폐: 객체는 자신의 상태를 캡슐화하고, 외부에서 직접 접근할 수 없도록 데이터를 은폐해야 합니다. 이를 통해 객체의 내부 상태를 보호하고 유지보수성을 향상시킵니다.

**Comment by GIVEN53** [comment]
1. 객체는 생명체이다. 객체는 속성(상태, 특징)과 행위(행동, 동작, 기능)을 가진다. 
2. 객체는 계약에 따라 동작한다. 인터페이스를 구현하는 객체라면, 좋은 객체의 모든 공용 메서드는 인터페이스에 선언된 메서드를 구현해야 한다.
3. 객체는 고유하다. 고유한 객체가 되기 위해서는 무언가를 캡슐화한 객체여야 한다.
4. 객체는 불변적이다. 좋은 객체는 캡슐화한 상태를 변하게 하지 않는다.
5. 객체의 클래스에 정적 멤버가 없다. 정적 메서드(static)은 객체의 행위가 아닌 클래스의 행위를 나타낸다. 
6. 객체에 실제의 유의미한 이름을 부여한다.
7. 모든 클래스는 final 아니면 abstract 클래스여야 한다.

**Comment by le2sky** [comment]
제가 생각하는 좋은 객체는 자율적이고, 협력적입니다.

협력적이라는 것은 다른 객체의 요청에 성실히 응답하고, 
필요한 경우에는 다른 객체에게도 적절히 도움을 요청할 수 있어야함을 의미합니다.
모든 것을 무시하고 스스로 해결하려는 객체는 god object가 되어 내부 복잡도에 의해 자멸할 것입니다.

자율적이라는 것은 클라이언트 객체의 요청에 대해 어떻게 응답할지 스스로 판단하고 결정하는 것을 의미합니다.
좋은 객체는 스스로 판단하고, 행동하는 자율적인 존재입니다.

**Reply by le2sky** [reply]
https://codingnuri.com/seven-virtues-of-good-object/

**Comment by GIVEN53** [comment]
자율 -> 스스로 판단, 행위
잘 숨겨짐 -> 캡슐화
